File: src/ctci/java/Ch1/Q5.java
package Ch1;

import java.util.Scanner;

/**
 * Created by Michael on 3/20/2017.
 *
 * Basic string compression  using counts of repeated adjacent characters
 * Return original string if new one is not smaller
 *
 * Takeaway: use char[] or stringbuffer to save computation on string concact
 */
public class Q5 {


    public static void main(String[] args) {
        System.out.println("Basic string compression: ");
        Scanner in = new Scanner(System.in);
        while (true) {
            System.out.print("Your input, separated by whitespace: ");
            String input = in.next();
            String result = basicCompression(input);
            if (input.length() <= result.length()) {
                System.out.println("result not smaller: " + result);
                System.out.println("original: " + input);
            } else {
                System.out.println("result is smaller: " + result);
            }
            System.out.println();

        }
    }

    public static String betterBasicCompression(String str) {
        int actualLengthCompression = -1; // TODO: do one pass throught he string to find what the length would be pre compress

        char[] result = new char[actualLengthCompression];
        int resultIndex = 0;
        char prevChar = str.charAt(0);
        int prevNumb = 0;
        for (int i = 0; i <= str.length(); i++) {
            if (i == str.length()) {
                result[resultIndex] = prevChar;
                result[resultIndex + 1] = (char) prevNumb;
            } else if (str.charAt(i) != prevChar) {
                result[resultIndex] = prevChar;
                result[resultIndex + 1] = (char) prevNumb;
                prevChar = str.charAt(i);
                prevNumb = 1;
                resultIndex += 2;
            } else {
                prevNumb++;
            }
        }
        return String.valueOf(result);
    }

    // assume we get a string of at least size 1
    public static String basicCompression(String str) {
        if (str.length() < 1) {
            return str;
        }

        String result = "";
        char prevChar = str.charAt(0);
        int prevNumb = 1;
        for (int i = 1; i < str.length(); i++) {
            if (str.charAt(i) != prevChar) {
                result += prevChar + "" + prevNumb;
                prevChar = str.charAt(i);
                prevNumb = 1;
            } else {
                prevNumb++;
            }
        }
        result += prevChar + "" + prevNumb;
        return result;
    }

}

File: src/ctci/python/Ch1/Q5.py
# Basic string compression  using counts of repeated adjacent characters
# Return original string if new one is not smaller
# TODO: optimize the string concatenation

def stringCompress(input):
    prevChar = input[0]
    prevNumb = 1
    result = ""

    for i in range(1, len(input)):
        if input[i] != prevChar: # output when reach differing adjacent characters
            result = result + prevChar + str(prevNumb)
            prevChar = input[i]
            prevNumb = 1
        else: # while reaching same adjacent, jsut keep track of numbers
            prevNumb += 1
    result += prevChar + str(prevNumb)
    return result




if __name__ == '__main__':

    print "Basic string compression  using counts of repeated adjacent characters"
    while True:
        print "Input string: "
        input = list(raw_input())
        result = stringCompress(input)
        if len(result) < len(input):
            print "Result is smaller: ", result
        else:
            print "Result not smaller: ", result
        print "result as string: ", ''.join(result)
        print